draft-jmiller-json-proof-algorithms.md

```diff
@@ -363,6 +365,59 @@ The same JWP in compact serialization:
 ImV5SnBjM01pT2lKb2RIUndjem92TDJsemMzVmxjaTVsZUdGdGNHeGxJaXdpWTJ4aGFXMXpJanBiSW1aaGJXbHNlVjl1WVcxbElpd2laMmwyWlc1ZmJtRnRaU0lzSW1WdFlXbHNJaXdpWVdkbElsMHNJblI1Y0NJNklrcFFWQ0lzSW1Gc1p5STZJa0pDVXkxWUluMCI.~IkpheSI~~NDI.AAUVqUnyMW_qGrrCzK7xHqUydeyQX5O6276HzUKz41aX9q7qzC6r-eHGh1GD4Y3ZL9l_sA9ibulK3OYbyXVy0ym3LEC5LYyZMUnHXdD41K9qx1P_gE2Y-qZlTvgKauwY7hUihoC6hJl7Vlic9ssX_fjZ6QGo0AzB3o5aHN_xgqSQUwAPpRmckfMnWPn2uZgD3Ei0AAAAdIW0OjVpeHzjoE7lStFvpGW3VnrUfloX5G24SUtysY6yWWtkvbtKj6fausG99w6GKAAAAAIFgnhW_xwhkRSVwwgEyNSyQ5PmrsAJnl6zkD88Uv_YBFVs0cLRsWpF1-uxU_QBw2aHYxR5B_wau84Q-oYB_8VcgmnxryMW0mWecdbeAwSovPsi1dPglpMmpOWX4t-8tbTM5DprN-iUtMcqtX876eq8AAAABF3EHyaXaecu5C5WxIjUr6pTFU4T4NN6B1xHEatCejiCLROs2UE_9icGHk2EViQpA-pIkshllp9sjA_DOfmav8xbHp8gnRgDblpmY84qOgG111UdIIMuTBkFK8sg16epbDiukjOuKL4WJV2c9dAaN7SHc0ZgavPVg8Q_pSqdFl4u
 ```
 
+
+## Salted HMAC
+
+The Salted HMAC (SH) algorithm is similar to Single Use (SU), but uses generated salts and HMAC instead of individual JWSs to protect each payload.  It requires less compute but can result in potentially larger presentation proof values.
+
+Like SU, it also does not support unlinkability if the same JWP is presented multiple times and requires an individually issued JWP for each presentation in order to fully protect privacy.
+
+### SU-based Setup
+
+This algorithm uses most of the same setup as the Single Use one, including the use of JWS as the signature mechanism and the holder providing a Presentation Key.
+
+The main difference with the SH algorithm is that instead of the issuer generating an Ephemeral Key for signing, it instead generates a Shared Secret using HMAC.  This Shared Secret is then used to generate unique salt values that become the input keys to HMAC the issuer protected header and each payload.
+
+### Issuer Protected Header
+
+The holder's Presentation Key JWK MUST be included in the issuer protected header using the `presentation_jwk` claim.  The issuer MUST validate that the holder has posession of this key through a trusted mechanism such as verifying the signature of a unique nonce value from the holder.
+
+The salt for the issuer header is generated using HMAC with the Shared Secret as the key and the value "issuer_header" as the value.  The issuer header JSON is then serialized using UTF-8 and encoded with base64url into an octet array.  The generated salt is used as the HMAC key with the octet array as the value, and the resulting hash value becomes the first input into the larger octet array that will be signed by the issuer.
```

Comment:
I wonder if we should not salt the issuer header instead just the payloads? as the header will never not be disclosed so no need to salt it?

Reply:
Also I'd refer to the HMAC key as a randomly generated value that is just shared with the holder and used to re-derive the required salts for payloads the holder wishes to disclose in a presentation

Reply:
I find the consistent treatment of the issuer header much more preferable from a security perspective.  The octet sequence that the issuer signs needs to be a uniform array-of-MACs so that it doesn't internally require any separator tags.

We could perhaps use the array of just the payload MACs as the key to MAC the issuer header, then the holder won't need to include its salt in the presentation proof and save another 32 or 64 bytes (depending on which MAC)?

Reply:
To me this is related to the conversation we have been having in https://github.com/decentralized-identity/bbs-signature/issues/102 which boils down to whether you want a cryptographically enforce-able seperation between information that is always required to be revealed vs information that is suppose to be selectively disclosed by the holder. If you opt for the former then it makes sense to treat the header differently at the cryptographic layer.

Reply:
The difference in my mind with JWP is that as a container spec it requires that the headers are always disclosed, the underlying algorithms don't need to provide the enforcement, JWP implementations will always do that.

It's more nuanced for BBS though as it can be used in many other applications without a spec to enforce it :)

Reply:
Yeah thats true, if you are intending on the crypto algorithm being used in a way that it is never expected to be used across an application context then I suppose this risk is mitigated, not entirely ruled out but mitigated. Although I still think handling the header differently (e.g just HASH rather than HMAC) would be worthwhile even just for reducing implementation complexity. With the current approach aren't you always required to reveal a salt/key for the header alongside any payloads?

Reply:
> Although I still think handling the header differently (e.g just HASH rather than HMAC) would be worthwhile even just for reducing implementation complexity. 

As a programmer I would agree, but with a standards hat on I definitely lean the other way: introducing another cryptographic primitive adds notable complexity to the spec, even if the MAC is based on a hash it's still a distinct primitive with different security assumptions.

> With the current approach aren't you always required to reveal a salt/key for the header alongside any payloads?

As currently defined yes, but we could eliminate that by using a fixed/spec-defined key to MAC the header, I can update the draft if that sounds reasonable?

Reply:
> As currently defined yes, but we could eliminate that by using a fixed/spec-defined key to MAC the header, I can update the draft if that sounds reasonable?

Yes I think I that would be a good improvement and strike a better balance! Keeps the scheme simple from a primative perspective but also doesn't induce the need to manage a random salt for the header where there is no need for it to be random